[DISCO-2373] Add Test Contribution Guidelines for Merino-py (#335)
mozilla-services/merino-py: CONTRIBUTING.md

Patches should be submitted as pull requests (PR).

Before submitting a PR:

- Ensure you are pulling from the most recent `main` branch and install dependencies with Poetry.
  See the [README][readme] for more details.
- Your code must run and pass all the automated tests before you submit your PR

>   for review. See the [README][readme] for information on linting, formatting and testing.

Reviewer: Nit: The README at https://github.com/mozilla-services/merino-py/blob/main/README.md looks pretty barren. 
Should we be directing people to the relevant details somewhere in the bowels of https://mozilla-services.github.io/merino-py/ instead?

Ah, I see you have a fancy link thing for `merino_book` below, but not sure how these fancy links work between sections, or if that is even the correct/best target. I guess ideally we'd be deep linking to the specific sections we're referring to.
```
[merino_book]: https://mozilla-services.github.io/merino-py/
```

Author: That's a good point! I've substituted the relevant page from the documentation.

  your reviewer's responsibility to ensure your patch includes adequate tests.
- If making changes that may impact performance, it is recommended to execute a load test. Please
  see the [Load Testing README][load_testing_readme] for more information.
- For more information on understanding the release process, please see relevant information
  contained in the [release-process.md][release_process] documentation.

[readme]: /README.md
[load_testing_readme]: /tests/load/README.md
[release_process]: /docs/dev/release-process.md

When submitting a PR:

- You agree to license your code under the project's open source license ([MPL 2.0][license]).
- Base your branch off the current `main`.
- Add both your code and new tests if relevant.
- [Sign][sign] your git commit.
- Run tests, linting and formatting checks to make sure your code complies with established standards.
(e.g. No warnings are returned for python: "`make lint`", "`make test`", "`make format`")
- Ensure your changes do not reduce code coverage of the test suite.
- Please do not include merge commits in pull requests; include only commits
  with the new relevant code.

> WIP: See the [merino book][merino_book] for information on prerequisites, installing, running and

Reviewer: What is the "WIP" here? Is this a TODO before you un-Draft and merge this PR?

Author: I think it stands for "Work In Progress" and the line must have persisted over time. We link to the merino book multiple times in this doc, I'm opting to remove the line.

* **chore**: Changes to the build process or auxiliary tools and libraries such as documentation
  generation

> Name the branch using this nomenclature with the `<type>` followed by a forward slash, followed by a
> dash-separated description of the task. Ex. `feat/add-sentry-sdk-MOZ-1234`. Note, if associated with
> a Jira ticket, synchronization with Jira and GitHub is possible by appending the suffix of the Jira
> ticket to the branch name (`-MOZ-1234` in the example above). This can be added anywhere in the
> branch name, but adding to the end is ideal. You can also include the Jira issue at the end of
> commit messages to keep the task up to date. See Jira Docs for referencing issues [here][jira]

Reviewer 2: Ah! I think we need to change this across the board, but I know we've typically been prepending the DISCO-1234 to the start of branch names, so maybe rewrite this and we can change this in the other services. So:

```suggestion
If associated with a Jira ticket, synchronization with Jira and GitHub is possible by appending the suffix of the Jiraticket to the branch name (`MOZ-1234` in the example below). Name the branch using this nomenclature with the optional `<type>` followed by a forward slash, followed by the Jira ticket and then a 
dash-separated description of the task. Ex. `feat/MOZ-1234-add-sentry-sdk` or `MOZ-1234-add-sentry-sdk` Note, the Jira ticket project and number can be added anywhere in the
branch name, but adding to the beginning is ideal. You can also include the Jira issue at the end of
commit messages to keep the task up to date. See Jira Docs for referencing issues [here][jira]
```

* **chore**: Changes to the build process or auxiliary tools and libraries such as documentation
  generation

Name the branch using this nomenclature with the `<type>` followed by a forward slash, followed by a

> dash-seperated description of the task. Ex. `feat/add-sentry-sdk-MOZ-1234`. Note, if associated with

Reviewer: Typo?
```suggestion
dash-separated description of the task. Ex. `feat/add-sentry-sdk-MOZ-1234`. Note, if associated with
```

Per https://www.dictionary.com/e/seperate-or-separate/
> _"Separated is the correct spelling, and seperated is a common misspelling. Like the word separate, the correct spelling has two a’s in it."_

a Jira ticket, synchronization with Jira and GitHub is possible by appending the suffix of the Jira
ticket to the branch name (`-MOZ-1234` in the example above). This can be added anywhere in the
branch name, but adding to the end is ideal. You can also include the Jira issue at the end of
commit messages to keep the task up to date. See Jira Docs for referencing issues [here][jira]

> [angular_commit_guidelines]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md

Reviewer: > This repository has been archived by the owner on Apr 12, 2022. It is now read-only.

nit: Link/guidelines are still valid, but not sure if there is a better place to link to.

Author: I don't have the context for this guideline, I'm not confident that I should modify it.

Reviewer: Ah, I vote we change it to the new Angular repo at https://github.com/angular/angular/blob/main/CONTRIBUTING.md

```diff
+ https://github.com/angular/angular/blob/main/CONTRIBUTING.md
- https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md
```

Author: Oh! Okay!

> Closes #314, #975

Reviewer: Unrelated to your PR, but I think this is subtly incorrect (if I understand GitHub Laws correctly).
This commit message would auto-close issue `#314` since it closes that. But the issue `#975` would NOT be auto-closed since you're merely mentioning it.
So this should be the slightly more verbose `Closes #314, Closes #975`
Ref: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

Author: I think you are right. I see comments on Stack Overflow indicating that this syntax 'used' to work. Correcting it!

      teardown methods. As test suites scale, setup and teardown methods become a common source of
      state issues and bloat. Tests will grow to have increasingly specific requirements that don't
      have common relevancy or compatibility.
* Tests should not rely on resources from sites whose content Mozilla doesn't control or where
  Mozilla has no SLA. This is a security concern as well as a source of intermittent failures.
* Tests should avoid file system or database dependencies, their changes or outages can be a source
  of intermittent failures.

**Repeatable**

* Tests should have consistent results when no code changes are made between test runs.
* Tests should not contain time bombs. Meaning they should not be susceptible to failure given
  execution during a given datetime or due to time comparisons.
* Tests should avoid using _magical_ time delays when waiting for operations to finish executing.
* Tests should not assume an execution order for asynchronous methods, this may lead to intermittent
  failures.
* Tests should not depend on objects through weak references, which may be garbage collected during
  test execution causing intermittent failures.

> * Tests should avoid using logical operations, such as if, while, for, and switch.

Reviewer: ```suggestion
* Tests should avoid using logical operations, such as `if`, `while`, `for`, and `switch`.
```

    * Decision points are a nexus for introducing bugs. Bugs found in test code erodes trust and
      diminishes value.
    * Consider splitting up a test or using a data driven test framework before using logical
      operations.
* Test CI/CD jobs should avoid pulling the latest language or tooling dependencies. New dependency
  versions can cause unintuitive failures.

**Self-Deterministic**

* No human intervention should be required to conclude if a test has passed or failed.

> **Timely**

Reviewer: TIL this is referred to as either "Timely" or "Thorough"
 - https://dzone.com/articles/first-principles-solid-rules-for-tests (both)
 - https://www.appsdeveloperblog.com/the-first-principle-in-unit-testing/ ("Thorough")
 - https://github.com/tekguard/Principles-of-Unit-Testing ("Thorough")
 - https://howtodoinjava.com/best-practices/first-principles-for-good-tests/ ("Timely")

Author: TIL too! `Thorough` is a good use of the `T`! 

In this case I'll prefer timely though, since the documented [test strategy](https://mozilla-services.github.io/merino-py/dev/testing.html) I think is a more specific guideline for what 'thorough' means in this project.

* Production code should be crafted to be testable, so that writing and maintaining tests doesn't
  take an unreasonable amount of time.

> * Test code should be writen with the same quality standard as production code and subject to the

Reviewer: typo
```suggestion
* Test code should be written with the same quality standard as production code and subject to the
```
